httpcore5/src/main/java/org/apache/hc/core5/http/protocol/RequestContent.java

```diff
@@ -110,6 +110,10 @@ public void process(final HttpRequest request, final EntityDetails entity, final
                 throw new ProtocolException("Content-Length header already present");
             }
         }
+        if (entity == null) {
```

Comment:
@chrysophylax I think this should be done for `POST`, `PUT` and `PATCH` only and not any other standard method. That would be consistent with the behavior of HC 4.x as well.

Reply:
I'd actually use a denyList with those methods that are documented not to anticipate content (which are only few).

Reply:
Isn't a HttpEntityEnclosingRequest inherently only for these methods which the standard defines Enclosed Content semantics?

Reply:
@chrysophylax We had to drop `HttpEntityEnclosingRequest` in 5.x because RFC 723x muddied the water by declaring it was legal for all methods to enclose an entity, well, except for `TRACE`, that is.

Reply:
RFC 723x did not "muddy the water" - it just documents how HTTP works in practice.

Reply:
Ahhh. I follow. 

@reschke are there more methods that will not anticipate content bar `TRACE`? I could modify it either way (only `POST`, `PUT`, `PATCH`) or a denyList (`TRACE`, `??`)

Reply:
@chrysophylax Please do `POST`, `PUT`, `PATCH` by default. If desired the users can always override the default behavior through a custom request interceptor.

Reply:
I believe TRACE is the only method that "MUST NOT" have content. There are some ("GET" and "DELETE"?) which "SHOULD NOT", so for these I wouldn't add "Content-Length: 0", either.